libcommon/tkey_debug.c

```diff
@@ -0,0 +1,162 @@
+// SPDX-FileCopyrightText: 2023 Tillitis AB <tillitis.se>
+// SPDX-License-Identifier: BSD-2-Clause
+
+#include <stdint.h>
+#include <tkey/lib.h>
+#include <tkey/proto.h>
+#include <tkey/tk1_mem.h>
+
+#ifndef TKEY_DEBUG
+#define TKEY_DEBUG
+#endif
+
+#include <tkey/debug.h>
+
+#define PACKET_SIZE 64
+#define HEADER_SIZE 2
+
+// clang-format off
+static volatile uint32_t *can_tx = (volatile uint32_t *)TK1_MMIO_UART_TX_STATUS;
+static volatile uint32_t *tx =     (volatile uint32_t *)TK1_MMIO_UART_TX_DATA;
+// clang-format on
+
+void tkey_writebyte(uint8_t b)
+{
+	for (;;) {
+		if (*can_tx) {
+			*tx = b;
+			return;
+		}
+	}
+}
+
+void tkey_write(const uint8_t *buf, size_t nbytes)
+{
+	for (int i = 0; i < nbytes; i++) {
+		tkey_writebyte(buf[i]);
+	}
+}
+
+void tkey_write_with_header(const uint8_t *buf, size_t nbytes)
+{
+	tkey_writebyte(MODE_TKEYCTRL);
+	tkey_writebyte(nbytes);
+
+	for (int i = 0; i < nbytes; i++) {
+		tkey_writebyte(buf[i]);
+	}
+}
+
+void tkey_putchar(const uint8_t ch)
+{
+	tkey_writebyte(MODE_TKEYCTRL);
+	tkey_writebyte(1);
+
+	tkey_writebyte(ch);
+}
+
+void tkey_lf()
+{
+	tkey_writebyte(MODE_TKEYCTRL);
+	tkey_writebyte(1);
+
+	tkey_writebyte('\n');
+}
+
+static char hexnibble(const uint8_t ch)
+{
+	switch (ch) {
+	case 0x0:
+		return '0';
+	case 0x1:
+		return '1';
+	case 0x2:
+		return '2';
+	case 0x3:
+		return '3';
+	case 0x4:
+		return '4';
+	case 0x5:
+		return '5';
+	case 0x6:
+		return '6';
+	case 0x7:
+		return '7';
+	case 0x8:
+		return '8';
+	case 0x9:
+		return '9';
+	case 0xa:
+		return 'a';
+	case 0xb:
+		return 'b';
+	case 0xc:
+		return 'c';
+	case 0xd:
+		return 'd';
+	case 0xe:
+		return 'e';
+	case 0xf:
+		return 'f';
+	}
+
+	return '0';
+}
+
+void tkey_puthex(const uint8_t ch)
+{
+	tkey_writebyte(MODE_TKEYCTRL);
+	tkey_writebyte(2);
+
+	tkey_writebyte(hexnibble(ch >> 4 & 0x0f));
+	tkey_writebyte(hexnibble(ch & 0x0f));
+}
+
+void tkey_putinthex(const uint32_t n)
+{
+	uint8_t buf[4];
+
+	tkey_writebyte(MODE_TKEYCTRL);
+	tkey_writebyte(2);
+
+	memcpy(buf, &n, 4);
+	tkey_write((uint8_t *)"0x", 2);
+	for (int i = 3; i > -1; i--) {
+		tkey_puthex(buf[i]);
+	}
+}
+
+void tkey_puts(const char *s)
+{
+	size_t remaining = strlen(s);
+
+	while (remaining > 0) {
+		uint8_t len = (remaining < (PACKET_SIZE - HEADER_SIZE))
```

Comment:
Here we split writes into chunks of 62 bytes.

If the intent here is to split the data into frames that fit in a USB HID packet we could split it into chunks of 64 bytes right? Something like the following

```
uint8_t len = (remaining < USB_PACKET_SIZE)
				  ? remaining
				  : USB_PACKET_SIZE;
```

where `USB_PACKET_SIZE` is 64